command/agent/http.go

```diff
@@ -501,6 +501,9 @@ func (s *HTTPServer) registerHandlers(enableDebug bool) {
 	s.mux.Handle("/v1/vars", wrapCORS(s.wrap(s.VariablesListRequest)))
 	s.mux.Handle("/v1/var/", wrapCORSWithAllowedMethods(s.wrap(s.VariableSpecificRequest), "HEAD", "GET", "PUT", "DELETE"))
 
+	// JWKS Handler
+	s.mux.HandleFunc("/.well-known/jwks.json", s.wrap(s.JWKSRequest))
```

Comment:
Typically the HTTP server will be protected with mTLS, which won't be reachable by third parties. So this probably needs to get served on a separate HTTP listener. (On its own port, I suppose?) This PR is already a nice size so we don't have to solve that in this PR, but just wanted to make sure we didn't forget about that problem.

Reply:
Yeah in the RFC (internal only, sorry folks) I call this out as a problem.

The Task API (+Workload Identity) gets around the mTLS problem for things running in Nomad... which our friends Consul and Vault usually are *not.*

...so then you're left with the option of running an mTLS terminating proxy just for this endpoint which is a hassle to say the least. How the proxy and the JWT's `issuer` field interacted could be problematic as well, although I don't think for Consul and Vault because their JWT auth methods allow you to specify the [JWKSURL explicitly](https://developer.hashicorp.com/consul/docs/security/acl/auth-methods/jwt#jwksurl) and only seem to use the [`issuer` claim for mapping](https://developer.hashicorp.com/consul/docs/security/acl/auth-methods/jwt#boundissuer).

That being said this is only a problem for folks running with [`tls.verify_https_client = true`](https://developer.hashicorp.com/nomad/docs/configuration/tls#verify_https_client) explicitly as it defaults to `false` even with mTLS enabled. I think this is still an ok default and recommendation as the HTTP endpoints will still be protected by (asymmetric) TLS, and the [Consul and Vault JWT auth methods support setting a custom CA cert for that.](https://developer.hashicorp.com/consul/docs/security/acl/auth-methods/jwt#jwkscacert)

.......even after all of that we should probably add a "Only TLS, Not mTLS" HTTP listener for this, the eventual OIDC discovery endpoint, and maybe other "low risk" endpoints like metrics.

While I feel like the situation has fairly reasonable options, I think it's clearly too complicated. Somebody hardening their cluster by setting `tls.verify_https_client = true` one day breaking their Consul integration due to the JWT auth method lacking a client certificate would just be hellish to debug. Definitely a : (╯°□°）╯︵ ┻━┻ situation I want to avoid.

Reply:
I keep forgetting about `tls.verify_https_client = false` being the default. Even [Vault's equivalent is set to false](https://developer.hashicorp.com/vault/docs/configuration/listener/tcp#tls_require_and_verify_client_cert).